### PR TITLE
rpc+lnrpc: add new closechannels cli+rpc command

### PR DESCRIPTION
**Pull request by joostjager**
https://github.com/lightningnetwork/lnd/issues/1250

### CODE REVIEW

**Comment by joostjager** [comment]
Created PR for an early review. Filter flags to be implemented.
I noticed that rpc generation returns some changes that aren't mine.

**Comment by joostjager** [comment]
Filtering added. I made it so that multiple filter flags can be specified (logical or)

File: .gitignore [hunk 1]
```diff
@@ -25,7 +25,9 @@ _testmain.go
 *.prof
 
 lnd
+lnd-debug
```
Comment (halseth) [inline .gitignore]: Thaaaaank you! 🤣

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -1520,6 +1520,58 @@ func listChannels(ctx *cli.Context) error {
...
+	Category: "Channels",
+	Usage:    "List all closed channels.",
+	Flags:    []cli.Flag{
+		cli.BoolFlag{
+			Name:  "cooperative",
+			Usage: "list channels that were closed cooperatively",
+		},
+		cli.BoolFlag{
+			Name:  "local_force",
+			Usage: "list channels that were force-closed by the local node",
```
Comment (halseth) [inline cmd/lncli/commands.go]: lines a bit too long. Try to wrap at 80 chars with 8 spaces indentation :)

File: lnrpc/rpc.proto [hunk 3]
```diff
@@ -854,6 +866,58 @@ message ListChannelsResponse {
...
+    /// Settled balance at the time of channel closure
+    int64 settled_balance = 8 [json_name = "settled_balance"];
+
+    /// The sum of all the time-locked outputs at the time of channel closure
+    int64 time_locked_balance = 9 [json_name = "time_locked_balance"];
+
+    enum ClosureType {
+        COOPERATIVE_CLOSE = 0;
+        LOCAL_FORCE_CLOSE = 1;
+        REMOTE_FORCE_CLOSE = 4;
```
Comment (halseth) [inline lnrpc/rpc.proto]: nit: I think we should either keep the ordering from `channeldb`, or the integer value, not both :D

File: rpcserver.go [hunk 4]
```diff
@@ -1570,6 +1574,77 @@ func (r *rpcServer) PendingChannels(ctx context.Context,
 	return resp, nil
 }
 
+// ClosedChannels returns a list of all the channels have been closed. This does not
```
Comment (halseth) [inline rpcserver.go]: line too long.

File: rpcserver.go [hunk 5]
```diff
@@ -1570,6 +1574,77 @@ func (r *rpcServer) PendingChannels(ctx context.Context,
...
+// include channels that are still in the process of closing.
+func (r *rpcServer) ClosedChannels(ctx context.Context,
+	in *lnrpc.ClosedChannelsRequest) (*lnrpc.ClosedChannelsResponse, error) {
+
+	// Show all channels when no filter flags are set
+	filterResults := in.Cooperative || in.LocalForce || in.RemoteForce || in.Breach || in.FundingCanceled
```
Comment (halseth) [inline rpcserver.go]: line too long. Goes for some other lines as well :)

File: rpcserver.go [hunk 6]
```diff
@@ -1570,6 +1574,77 @@ func (r *rpcServer) PendingChannels(ctx context.Context,
 	return resp, nil
 }
 
+// ClosedChannels returns a list of all the channels have been closed. This does not
+// include channels that are still in the process of closing.
+func (r *rpcServer) ClosedChannels(ctx context.Context,
+	in *lnrpc.ClosedChannelsRequest) (*lnrpc.ClosedChannelsResponse, error) {
+
+	// Show all channels when no filter flags are set
```
Comment (halseth) [inline rpcserver.go]: nit: period at the end of sentences.

**Comment by joostjager** [comment]
@halseth Thanks for reviewing. Changes made.

**Comment by Roasbeef** [comment]
Needs a rebase to address proto conflicts.

File: .gitignore (hunk 1 above)
Comment (Roasbeef) [reply .gitignore]: 👍

File: lnrpc/rpc.proto [hunk 7]
```diff
@@ -854,6 +866,58 @@ message ListChannelsResponse {
...
+message ClosedChannelsRequest {
+    bool cooperative = 1;
+    bool local_force = 2;
+    bool remote_force = 3;
+    bool breach = 4;
+    bool funding_canceled = 5;
+}
+
+message ClosedChannelsResponse { 
+    repeated ChannelCloseSummary channels = 11 [json_name = "channels"];
```
Comment (Roasbeef) [inline lnrpc/rpc.proto]: Why start the numbering at 11 here?

Comment (joostjager) [reply lnrpc/rpc.proto]: Copied from ListChannels call. Changed to id 1.

**Comment by joostjager** [comment]
@Roasbeef Change made and rebased.

**Comment by halseth** [comment]
Yeah, that happens. I'll just approve again :)

**Pull request merged**